### ISSUE TITLE
Model does not work

### ISSUE THREAD

**Issue by AidasK** [yingcong/Facelet_Bank#7]
Hi, I am using my own trained models, but input and result is almost the same. Result is a bit brighter. I have tried to train for 5 minutes and for 24 hours. In both cases `loss_all` drops down to 0.310 ~ 0.211. I tried training with 5000 attribute vectors and with 50. No difference. Maybe there is a way to debug attribute_vector/ generated npz files? 

@yingcong Have you encountered such issue?

**Comment by yingcong** [comment]
Hi, @AidasK 
One useful tricks for debugging is to visualize its heatmap (Fig. 3 and Fig. 4 in the paper). If  the attribute vector is correct, then it should have large activation on the related regions. Thus to debug your system, you can visualize both npz file and the learned facelet output to locate the problem. 

The implemetation is described as follows.
Let w1 denote the attribute vector of one certain layer whose shape is nSamples x nChannels x Height x Width. we can use   
cv2.imwrite('heatmap1.jpg', torch.sum(w1[n]*w1[n], dim=0))
to visualize the heatmap of the n-th sample.

**Comment by AidasK** [comment]
Thanks for your quick and detailed response as always. I was able to generate a heatmap for facehair attribute for each image like this:

```python
def heatmap(flattened_features, name=''):
  flattened_features = torch.from_numpy(flattened_features)
  fn1 = flattened_features[: 256 * 112 * 112].view(256, 112, 112).numpy()
  fn2 = flattened_features[256 * 112 * 112: 256 * 112 * 112 + 512 * 56 * 56].view(512, 56, 56).numpy()
  cv2.imwrite('heatmap/%s.jpg' % name, numpy.sum(fn1, axis=0))
  # cv2.imwrite('heatmap/56%s.jpg' % name, numpy.average(fn2, axis=0))
  print('heatmap/112%s.jpg' % name)
```
And in demo2_facelet.py:
```python
        PF = model.mean_F(utils.warped_image_feed(xP, MP[idxP[:K]], image_dims))
        QF = model.mean_F(utils.warped_image_feed(xQ, MQ[idxQ[:K]], image_dims))
        WF = (QF - PF)
        shutil.copy(xX, 'heatmap/' + os.path.basename(xX))
        deepmodels_torch.heatmap(PF, os.path.basename(xX) + 'n')
        deepmodels_torch.heatmap(QF, os.path.basename(xX) + 'y')
        deepmodels_torch.heatmap(WF, os.path.basename(xX) + 'diff')
```
028896.jpg heatmap:
![028896](https://user-images.githubusercontent.com/2088484/42566499-78b6ff3e-850f-11e8-81a1-2cb196c68aa6.jpg)
WF:
![028896 jpgdiff](https://user-images.githubusercontent.com/2088484/42566501-78d01cbc-850f-11e8-8fa5-87d47a979b5c.jpg)
PF:
![028896 jpgn](https://user-images.githubusercontent.com/2088484/42566503-78edc3e8-850f-11e8-8351-2d3aec79ed36.jpg)
QF:
![028896 jpgy](https://user-images.githubusercontent.com/2088484/42566504-790a0a44-850f-11e8-88f2-d0a0875213df.jpg)


020690.jpg heatmap:
![020690](https://user-images.githubusercontent.com/2088484/42567020-26b018c2-8511-11e8-8b81-937df2cbc90b.jpg)
WF:
![020690 jpgdiff](https://user-images.githubusercontent.com/2088484/42567021-26cd0630-8511-11e8-9dee-18f3e2a1ea66.jpg)
PF:
![020690 jpgn](https://user-images.githubusercontent.com/2088484/42567022-26e81a7e-8511-11e8-9b6b-88a7f3a47c04.jpg)
QF:
![020690 jpgy](https://user-images.githubusercontent.com/2088484/42567023-2705419e-8511-11e8-8bf0-bb5782450ffd.jpg)


I can see that moustache are a bit highlighted, but everything around are too. I think my problem is that dataset variance is too high. Heatmap is built from faces which are looking in different directions and thus attributes don't line up. 

Also I cut corners with `facemodel_attributes.score()` method, by returning attribute scores in binary format 1,0 (which are taken from list_attr_celeba.txt csv file). This might have affected results too, but I don't know. 

Please let me know if this sounds logical to you. Thanks! :)

**Comment by yingcong** [comment]
Indeed, if the dataset has a very large variance then the training could be much more difficult (especially for the second case). In my experiments, I also find that this approach does not do quite well in non-frontal faces. 

But it seems not quite reasonable that it learns nothing. I was wondering if 1. the effect is not strong enough to be noticed; 2. the decoder is not adaptable to the new dataset. Thus my suggestions are 1. try a very large edit strength (e.g, 20) to see if it really learn something; 2. replace the decoder with backward based method to reverse the feature back to the image. The code is illustrated as bellow. 
Please extend the the VGG class https://github.com/yingcong/Facelet_Bank/blob/0f2edb49ea72b4da5da6a12d08cbcd4ff7811f8f/network/base_network.py#L84
with the following method. 

```python
    def backward_recon(self, img, target_feature, lr=1, iter=500, tv=1e1):
        '''
        backward to reconstruct the target feature
        :param img: the input image
        :param target_feature: the target feature, it is a list whose length is 3
        :return: output image
        '''
        _img = util.toVariable(img.cuda(), requires_grad=True)
        target_feature = [util.toVariable(t.cuda(), requires_grad=False) for t in target_feature]

        optim = torch.optim.LBFGS([_img], lr=lr, max_iter=iter)
        optim.n_steps = 0

        c1 = torch.nn.MSELoss(size_average=False).cuda()
        c2 = torch.nn.MSELoss(size_average=False).cuda()
        c3 = torch.nn.MSELoss(size_average=False).cuda()
        tv_loss = TVLoss()

        def step():
            self.features_1.zero_grad()
            self.features_2.zero_grad()
            self.features_3.zero_grad()
            if _img.grad is not None:
                _img.grad.data.fill_(0)
            feat = self.forward(_img)
            loss1 = c1(feat[0], target_feature[0])
            loss2 = c2(feat[1], target_feature[1])
            loss3 = c3(feat[2], target_feature[2])
            losstv = tv_loss(_img)
            loss = loss1 + loss2 + loss3 + tv * losstv
            loss.backward()
            if optim.n_steps % 25 == 0:
                print('%d, loss3=%f,loss4=%f,loss5=%f,TVloss=%f max=%d' % (
                    optim.n_steps, loss1.data[0], loss2.data[0], loss3.data[0], losstv.data[0], iter))
            optim.n_steps += 1
            return loss

        # _img.data.sub_(_img.grad.data * lr)
        optim.step(step)
        # print(_img.grad.data)
        img = _img.cuda()
        return img

class TVLoss(nn.Module):
    def __init__(self, eps=1e-3, beta=2):
        super(TVLoss2, self).__init__()
        self.eps = eps
        self.beta = beta

    def forward(self, input):
        x_diff = input[:, :, :-1, :-1] - input[:, :, :-1, 1:]
        y_diff = input[:, :, :-1, :-1] - input[:, :, 1:, :-1]

        sq_diff = torch.clamp(x_diff * x_diff + y_diff * y_diff, self.eps, 10000000)
        return torch.norm(sq_diff, self.beta / 2.0) ** (self.beta / 2.0)
```
This maps the shifted feature back to the image in a more reliable (but slower) way, so that you can use it to debug your code.

**Comment by AidasK** [comment]
I have tried with strength 4,5,9, and 18000. Face does not change with my trained model. With yours, it does.

backward_recon did not work as is, so I made one adjustment: 
```py
            loss1 = c1(feat[0], target_feature[0])
            loss2 = c2(feat[1], target_feature[1])
            loss3 = c3(feat[2], target_feature[2])
```
changed to 
```py
loss1 = c1(feat[0][0], target_feature[0])
loss2 = c2(feat[1][0], target_feature[1])
loss3 = c3(feat[2][0], target_feature[2])
```

And full usage is in modified test_facelet_net.py:
```python
def test_image():
    parser = argparse.ArgumentParser()
    parser.add_argument('command', help='test_image')
    parser.add_argument('-a', '--attribute', default='', help='attribute_vector/202384_facehair.npz')
    parser.add_argument('-i', '--input', default='examples/input.png',
                        help='path of the testing image. Use comma to separate each path. If this argument is a directory, then it will test all images in this directory.')
    args = parser.parse_args()
    vgg = VGG()
    args.pretrained = not args.local_model
    facelet = Facelet(args)
    if args.local_model:
        facelet.load(args.effect, pretrain_path='checkpoints')
    decoder = vgg_decoder()
    if not args.cpu:
        vgg = vgg.cuda()
        facelet = facelet.cuda()
        decoder = decoder.cuda()
    if os.path.isdir(args.input):
        print('input path is a directory. All images in this folder will be tested.')
        image_list = glob.glob(args.input + '/*')
    else:
        image_list = args.input.split(',')
    dataset = Dataset(image_list, scale=util.str2numlist(args.size))
    dataloader = DataLoader(dataset, batch_size=1, shuffle=False, num_workers=1)
    for idx, data in enumerate(tqdm(dataloader), 0):
        image, filename, image_shape = data
        filename = filename[0]
        if args.attribute:
            av = np.load(args.attribute)['WF']
            av = vec2featmap(av, 448)
            output = vgg.backward_recon(image, av)
            output = untransform(output.data[0].cpu())
        imageio.imwrite('%s-%s-s-%d.%s' % (util.remove_format_name(filename), args.effect, args.strength, filename[-3:]),
                        output)
```
And results are:
```
$ python test_facelet_net.py test_image -i examples/009296.jpg --effect facehair --strength 18000 -a attribute_vector/009296_facehair.npz
loading pretrained weights of VGG encoder
loading pretrained weights of VGG decoder
* Total Images: 1
  0%|                                                           | 0/1 [00:00<?, ?it/s]/home/aidas/quiz-facebank/network/base_network.py:170: UserWarning: invalid index of a 0-dim tensor. This will be an error in PyTorch 0.5. Use tensor.item() to convert a 0-dim tensor to a Python number
  optim.n_steps, loss1.data[0], loss2.data[0], loss3.data[0], losstv.data[0], iter))
0, loss3=10617220.000000,loss4=4555681.000000,loss5=3899012.500000,TVloss=63000.937500 max=500
25, loss3=2183703.750000,loss4=820561.562500,loss5=188771.234375,TVloss=18009.921875 max=500
50, loss3=1366472.875000,loss4=532925.687500,loss5=129431.085938,TVloss=9867.266602 max=500
75, loss3=966228.625000,loss4=380957.406250,loss5=93735.140625,TVloss=7289.816895 max=500
100, loss3=670099.562500,loss4=259576.312500,loss5=72812.531250,TVloss=5241.864258 max=500
125, loss3=460194.062500,loss4=174063.328125,loss5=60899.132812,TVloss=3364.117676 max=500
150, loss3=331807.906250,loss4=120935.351562,loss5=53118.062500,TVloss=2219.300293 max=500
175, loss3=241670.656250,loss4=85155.140625,loss5=47482.859375,TVloss=1680.104736 max=500
200, loss3=191445.093750,loss4=65809.382812,loss5=43481.152344,TVloss=1180.192139 max=500
225, loss3=166282.453125,loss4=56498.210938,loss5=41335.304688,TVloss=910.969910 max=500
250, loss3=152737.578125,loss4=51561.582031,loss5=40174.515625,TVloss=801.345947 max=500
275, loss3=144805.875000,loss4=48752.125000,loss5=39478.257812,TVloss=754.785278 max=500
300, loss3=139688.015625,loss4=46875.945312,loss5=39000.382812,TVloss=731.841980 max=500
325, loss3=136189.968750,loss4=45623.191406,loss5=38713.933594,TVloss=722.155396 max=500
350, loss3=133666.921875,loss4=44720.425781,loss5=38516.824219,TVloss=713.064575 max=500
375, loss3=131973.468750,loss4=44092.402344,loss5=38367.398438,TVloss=706.141602 max=500
400, loss3=130801.695312,loss4=43673.234375,loss5=38256.339844,TVloss=701.071045 max=500
425, loss3=129954.523438,loss4=43381.667969,loss5=38179.945312,TVloss=698.217773 max=500
450, loss3=129317.906250,loss4=43158.890625,loss5=38124.117188,TVloss=696.289124 max=500
475, loss3=128829.390625,loss4=42987.039062,loss5=38078.589844,TVloss=695.146790 max=500
100%|██████████████████████████████████████████████████| 1/1 [13:56<00:00, 836.74s/it]
```

![009296](https://user-images.githubusercontent.com/2088484/42894904-d96d09a2-8ac0-11e8-8569-0201227e2b93.jpg)
![009296-facehair-s-18000](https://user-images.githubusercontent.com/2088484/42894905-d98b293c-8ac0-11e8-8096-fea04b3b7fa5.jpg)
I can definitely see some beard, but this is not the result I expected. Probably I have used it wrong. It's hard to debug this without clear understanding.

**Comment by MdAshrafulAlam** [comment]
Where did you get this "202384_facehair.npz"?

**Comment by AidasK** [comment]
This is an attribute vector made by demo2_facelet.py code

**Comment by MdAshrafulAlam** [comment]
Could you please tell me how to get Heat Map Image?
